python/python/bystro/_template_npr.py

```diff
@@ -0,0 +1,121 @@
+"""
+This provides a basic template for any model that uses numpyro as an 
+inference method. It has several methods that should be filled by any
+object extending the template, namely
+    fit
+    _fill_hp_options
+
+These are the methods for running the samples given data and providing 
+hyperameter selections respectively.
+
+Objects
+-------
+_BaseNPRModel(object)
+    This is the template Numpyro model
+
+"""
+import abc
+import cloudpickle # type: ignore
+import numpyro # type: ignore
+from copy import deepcopy
+
+
+class _BaseNPRModel(object):
+    def __init__(self, mcmc_options={}, hp_options={}):
+        """
+
+        Parameters
+        ----------
+
+        Returns
+        -------
+
+        """
+        self.mcmc_options = self._fill_mcmc_options(mcmc_options)
+        self.hp_options = self._fill_hp_options(hp_options)
+        self.samples = None
+
+    def fit(self, *args):
+        """
+
+        Parameters
+        ----------
+
+        Returns
+        -------
+
+        """
+
+    def render_model(self):
+        """
+
+        Parameters
+        ----------
+
+        Returns
+        -------
+
+        """
+        assert self._model is not None, "Fit model first"
```

Comment:
@poneill what do you prefer for user-facing errors in situations like this? Pep suggests RuntimeError afaik?

Reply:
Yes, I left a comment about this too but it looks like github ate it.  

In general we should code as if our asserts could be turned off at any time, because they can.  That means that asserts are basically only good for catching errors that should *never* happen (as an aid to development / debugging) and communicating invariants to the reader.  If we have to check for an error that could happen for reasons beyond the code's control, such as ValueErrors or a user using a class incorrectly, we should check explicitly and raise the appropriate exception (here, a RuntimeError).